src/lint.cpp

```diff
@@ -280,11 +280,23 @@ void linter::visit_exit_for_scope() {
   this->scopes_.pop();
 }
 
+[[nodiscard]] bool linter::is_eval_used() {
+  scope &current_scope = this->current_scope();
+  for (const used_variable &used_var : current_scope.variables_used) {
+    if (used_var.name.normalized_name() == u8"eval") {
+      return true;
+    }
+  }
+  return false;
+}
+
 void linter::visit_exit_function_scope() {
   QLJS_ASSERT(!this->scopes_.empty());
-  this->propagate_variable_uses_to_parent_scope(
-      /*allow_variable_use_before_declaration=*/true,
-      /*consume_arguments=*/true);
+  if (!is_eval_used()) {
```

Comment:
I think this approach will cause false negatives for code like the following:
```js
let eval = console.log;
function g() {
  eval("hello");
  undeclaredVariable;
}
g();
```
False negatives (i.e. not reporting an error) are better than false positives (i.e. reporting a bogus error), so I'm fine with this tradeoff if we can't find a better solution.

Reply:
I have tried to come up with some logic to handle this, let me know if it fails in any other condition.

Reply:
I just realized that it has a bug in the following code:

```js
function f() {
  eval("var x = 42;")
}
f();

function g() {
  undeclaredVariable;    // quick-lint-js should report an error here but it doesn't
}
```

I guess we will have to make a trade-off here, with the current system it doesn't seem that we can handle all the cases.